geopandas/io/sql.py

```diff
@@ -96,3 +96,222 @@ def load_geom_text(x):
+        else:
+            basic_types.append(gt)
+    geom_types = list(set(basic_types))
+
+    # Check for mixed geometry types
+    assert len(geom_types) < 2, "GeoDataFrame contains mixed geometry types."
+    geom_type = geom_types[0]
+    return (geom_type, geom_collection)
+
+
+def get_srid_from_crs(gdf):
+    """
+    Get EPSG code from CRS if available. If not, return -1.
+    """
+    from pyproj import CRS
+
+    if gdf.crs is not None:
+        try:
+            if isinstance(gdf.crs, dict):
+                # If CRS is in dictionary format use only the value
+                # to avoid pyproj Future warning
+                if "init" in gdf.crs.keys():
+                    srid = CRS(gdf.crs["init"]).to_epsg(min_confidence=25)
+                else:
+                    srid = CRS(gdf.crs).to_epsg(min_confidence=25)
+            else:
+                srid = CRS(gdf).to_epsg(min_confidence=25)
+            if srid is None:
+                srid = -1
+        except Exception:
+            srid = -1
+            print(
+                "Warning: Could not parse CRS from the GeoDataFrame.",
+                "Inserting data without defined CRS.",
+            )
+    return srid
+
+
+def convert_to_wkb(gdf, geom_name):
+    """Convert geometries to wkb. """
+    from shapely.wkb import dumps
+
+    gdf[geom_name] = gdf[geom_name].apply(lambda x: dumps(x, hex=True))
+    return gdf
+
+
+def write_to_db(gdf, engine, index, tbl, srid, geom_name, if_exists):
+    import io
+    import csv
+
+    # Convert columns to lists and make a generator
+    args = [list(gdf[i]) for i in gdf.columns]
+    if index:
+        args.insert(0, list(gdf.index))
+
+    data_iter = zip(*args)
+
+    # get list of columns using pandas
+    keys = tbl.insert_data()[0]
+    columns = ", ".join('"{}"'.format(k) for k in list(keys))
+
+    s_buf = io.StringIO()
+    writer = csv.writer(s_buf)
+    writer.writerows(data_iter)
+    s_buf.seek(0)
+
+    conn = engine.raw_connection()
+    cur = conn.cursor()
+
+    try:
+        # If appending to an existing table, temporarily change
+        # the srid to 0, and update the SRID afterwards
+        if if_exists == "append":
+            sql = "SELECT UpdateGeometrySRID('{schema}','{tbl}','{geom}',{crs})".format(
+                schema=tbl.table.schema, tbl=tbl.table.name, geom=geom_name, crs=0
+            )
+            cur.execute(sql)
+
+        sql = "COPY {} ({}) FROM STDIN WITH CSV".format(tbl.table.fullname, columns)
+        cur.copy_expert(sql=sql, file=s_buf)
+
+        # SRID needs to be updated afterwards as the current approach does not support
+        # the use of EWKT geometries.
+        sql = "SELECT UpdateGeometrySRID('{schema}','{tbl}','{geom}',{srid})".format(
+            schema=tbl.table.schema, tbl=tbl.table.name, geom=geom_name, srid=srid
+        )
+        cur.execute(sql)
+        conn.commit()
+
+    except Exception as e:
+        conn.connection.rollback()
+        raise e
```

Comment:
Do you know which Exception is expected? Can we provide some error message here to describe what happened?

Reply:
There might be different errors happening when pushing data to the database. The error messages raised by psycopg2 are typically quite good, so I maybe would not deal them separately here.

Reply:
Alright, makes sense.